### PR TITLE
Fixes #443: Add new ResetReason enum items

### LINKED ISSUE #443: ResetReason not dealing with all reset alternatives panics when reason is esp_reset_reason_t_ESP_RST_USB
On reset reason translation from IDF code to ResetReason struct at https://github.com/esp-rs/esp-idf-hal/blob/master/src/reset.rs#L34 the code panics if the reset reason is not in a subset of all reasons. In the bindings.rs generated from esp-idf-sys version "0.34.1" the reasons are:

```rust
#[doc = "!< Reset reason can not be determined"]
pub const esp_reset_reason_t_ESP_RST_UNKNOWN: esp_reset_reason_t = 0;
#[doc = "!< Reset due to power-on event"]
pub const esp_reset_reason_t_ESP_RST_POWERON: esp_reset_reason_t = 1;
#[doc = "!< Reset by external pin (not applicable for ESP32)"]
pub const esp_reset_reason_t_ESP_RST_EXT: esp_reset_reason_t = 2;
#[doc = "!< Software reset via esp_restart"]
pub const esp_reset_reason_t_ESP_RST_SW: esp_reset_reason_t = 3;
#[doc = "!< Software reset due to exception/panic"]
pub const esp_reset_reason_t_ESP_RST_PANIC: esp_reset_reason_t = 4;
#[doc = "!< Reset (software or hardware) due to interrupt watchdog"]
pub const esp_reset_reason_t_ESP_RST_INT_WDT: esp_reset_reason_t = 5;
#[doc = "!< Reset due to task watchdog"]
pub const esp_reset_reason_t_ESP_RST_TASK_WDT: esp_reset_reason_t = 6;
#[doc = "!< Reset due to other watchdogs"]
pub const esp_reset_reason_t_ESP_RST_WDT: esp_reset_reason_t = 7;
#[doc = "!< Reset after exiting deep sleep mode"]
pub const esp_reset_reason_t_ESP_RST_DEEPSLEEP: esp_reset_reason_t = 8;
#[doc = "!< Brownout reset (software or hardware)"]
pub const esp_reset_reason_t_ESP_RST_BROWNOUT: esp_reset_reason_t = 9;
#[doc = "!< Reset over SDIO"]
pub const esp_reset_reason_t_ESP_RST_SDIO: esp_reset_reason_t = 10;
#[doc = "!< Reset by USB peripheral"]
pub const esp_reset_reason_t_ESP_RST_USB: esp_reset_reason_t = 11;
#[doc = "!< Reset by JTAG"]
pub const esp_reset_reason_t_ESP_RST_JTAG: esp_reset_reason_t = 12;
```

Reasons with code greater than 10 are not included in the translation.

I have a board with an ESP32c3 module that on first boot after flashing it panics. The code is this (https://gitlab.com/scrobotics/optical-makerspace/dark-sky-meter-fw/-/blob/104-app-panics-just-on-first-boot/examples/panics_on_reset_reason.rs?ref_type=heads):
```rust
// This example demonstrates a panic that occurs when trying to get the reset reason.
use std::{thread, time};

use esp_idf_sys::{self as _, esp_reset_reason}; // If using the `binstart` feature of `esp-idf-sys`, always keep this module imported


fn main() {
    esp_idf_sys::link_patches();

    let reset_reason_code = unsafe { esp_reset_reason() };
    println!("Reset reason code: {:?}", reset_reason_code);

    let reset_reason = esp_idf_hal::reset::ResetReason::get(); // this panics on first run
    println!("Reset reason: {:?}", reset_reason);

    loop {
        thread::sleep(time::Duration::from_millis(1000)); // So the background task does not starve
    }
}

```

the output after flashing is:

```
I (24) boot: ESP-IDF v5.1.2-342-gbcf1645e44 2nd stage bootloader
I (24) boot: compile time Dec 12 2023 10:50:58
I (25) boot: chip revision: v0.4
I (29) boot.esp32c3: SPI Speed      : 40MHz
I (34) boot.esp32c3: SPI Mode       : DIO
I (38) boot.esp32c3: SPI Flash Size : 4MB
[...]

I (436) main_task: Started on CPU0
I (436) main_task: Calling app_main()
Reset reason code: 11
thread 'main' panicked at /home/jorgem/.cargo/registry/src/index.crates.io-6f17d22bba15001f/esp-idf-hal-0.43.1/src/reset.rs:47:18:
internal error: entered unreachable code
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

abort() was called at PC 0x420234a3 on core 0
0x420234a3 - panic_abort::__rust_start_panic::abort
    at /home/jorgem/.rustup/toolchains/nightly-2024-02-01-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/panic_abort/src/lib.rs:48
[...]

I (24) boot: ESP-IDF v5.1.2-342-gbcf1645e44 2nd stage bootloader
I (24) boot: compile time Dec 12 2023 10:50:58
I (24) boot: chip revision: v0.4
I (28) boot.esp32c3: SPI Speed      : 40MHz
I (33) boot.esp32c3: SPI Mode       : DIO
I (38) boot.esp32c3: SPI Flash Size : 4MB
[...]

I (436) main_task: Started on CPU0
I (436) main_task: Calling app_main()
Reset reason code: 4
Reset reason: Panic

```

I have also tested with an ESP32-C3-DevKit and the reset reason is different (maybe that is the reason because it has not been issued earlier):
```
I (32) boot: ESP-IDF v5.1.2-342-gbcf1645e44 2nd stage bootloader
I (32) boot: compile time Dec 12 2023 10:50:58
I (32) boot: chip revision: v0.3
[...]

I (444) main_task: Calling app_main()
Reset reason code: 1
Reset reason: PowerOn

```

### PR DESCRIPTION
**Pull request by punkto**

### CODE REVIEW

**Comment by punkto** [comment]
Closes https://github.com/esp-rs/esp-idf-hal/issues/443

File: src/reset.rs [hunk 1]
```diff
@@ -27,6 +27,51 @@ pub enum ResetReason {
     TaskWatchdog,
     /// Reset after exiting deep sleep mode
     DeepSleep,
+    /// Reset by USB peripheral (introduced in IDF v5.1.4)
+    #[cfg(not(any(
+        esp_idf_version_major = "4",
+        all(esp_idf_version_major = "5", esp_idf_version_minor = "0"),
+        esp_idf_version_full = "5.1.0",
+        esp_idf_version_full = "5.1.1",
+        esp_idf_version_full = "5.1.2",
```
Comment (Vollbrecht) [inline src/reset.rs:36]: missing 5.1.3 case. Not nice that they made that change on a patch release and not a minor release ;D

File: src/reset.rs [hunk 2]
```diff
@@ -27,6 +27,51 @@ pub enum ResetReason {
...
+    )))]
+    USBPeripheral,
+    /// Reset by JTAG (introduced in IDF v5.1.4)
+    #[cfg(not(any(
+        esp_idf_version_major = "4",
+        all(esp_idf_version_major = "5", esp_idf_version_minor = "0"),
+        esp_idf_version_full = "5.1.0",
+        esp_idf_version_full = "5.1.1",
+        esp_idf_version_full = "5.1.2",
+    )))]
```
Comment (Vollbrecht) [inline src/reset.rs:46]: dito 5.1.3 case

**Review by Vollbrecht — CHANGES_REQUESTED** [review]
Besides the minor forgotten case the PR looks good.

Not nice that they introduced the changes in a minor patch release, but yeah in C adding a enum variant is not a breaking change ;D

File: src/reset.rs (hunk 1 above)
Comment (punkto) [reply src/reset.rs:36]: :facepalm:

**Review by Vollbrecht — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**